### PR TITLE
Maximal square

### PR DESCRIPTION
**Pull request by fbokovikov**

### CODE REVIEW

File: src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    val m = matrix[0].size
+    val dp = Array(n) { IntArray(m) }
+    for (i in 0 until n) dp[i][0] = decode(i, 0)
+    for (j in 0 until m) dp[0][j] = decode(0, j)
+    for (i in 1 until n)
+        for (j in 1 until m)
+            if (matrix[i][j] == '1')
+                dp[i][j] = minOf(dp[i - 1][j], dp[i - 1][j - 1], dp[i][j - 1]) + 1
+            else
+                dp[i][j] = 0
```
Comment (DmitryBokovikov) [inline src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt:21]: don't need this branch

File: src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package leetcode.challenge.april.week4
+
+/**
+ * Given a 2D binary matrix filled with 0's and 1's, find the largest square containing only 1's and return its area.
+ */
+fun maximalSquare(matrix: Array<CharArray>): Int {
+    if (matrix.isEmpty()) {
+        return 0
+    }
```
Comment (DmitryBokovikov) [inline src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt:9]: don't need this check. It's not special case, your alg should support it.

File: src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    fun decode(i: Int, j: Int) = if (matrix[i][j] == '1') 1 else 0
```
Comment (DmitryBokovikov) [inline src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt:10]: you like overengineering bro, don't you?

File: src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    for (i in 1 until n)
+        for (j in 1 until m)
+            if (matrix[i][j] == '1')
+                dp[i][j] = minOf(dp[i - 1][j], dp[i - 1][j - 1], dp[i][j - 1]) + 1
+            else
+                dp[i][j] = 0
+    var max = 0
+    for (i in 0 until n)
+        for (j in 0 until m)
+            max = maxOf(max, dp[i][j])
```
Comment (DmitryBokovikov) [inline src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt:25]: don't need this branch, sustain max value during dp calculation

File: src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+fun maximalSquare(matrix: Array<CharArray>): Int {
+    if (matrix.isEmpty()) {
+        return 0
+    }
+    fun decode(i: Int, j: Int) = if (matrix[i][j] == '1') 1 else 0
+    val n = matrix.size
+    val m = matrix[0].size
+    val dp = Array(n) { IntArray(m) }
+    for (i in 0 until n) dp[i][0] = decode(i, 0)
+    for (j in 0 until m) dp[0][j] = decode(0, j)
```
Comment (DmitryBokovikov) [inline src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt:15]: don't need this and previous lines, extend boarders of dp table

**Comment by DmitryBokovikov** [comment]
Please, pay attention how I implemented the same problem

File: src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt (hunk 1 above)
Comment (fbokovikov) [reply src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt]: That's true

File: src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt (hunk 2 above)
Comment (fbokovikov) [reply src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt]: Basically, it's depends on your preferences

File: src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt (hunk 3 above)
Comment (fbokovikov) [reply src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt]: Have ever heard about DRY?

File: src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt (hunk 4 above)
Comment (fbokovikov) [reply src/main/kotlin/leetcode/challenge/april/week4/MaximalSquare.kt]: That once again true

**Pull request merged**